### DISCUSSION TITLE
What does globalCompFraction mean in GEOSX?

### DISCUSSION THREAD

**Discussion by wdmhouston** [GEOS-DEV/GEOS#1457] [General]
Hi expert,
 
    Please find a small CO2 injection test  example:
[geosx_co2mass.zip](https://github.com/GEOSX/GEOSX/files/6632449/geosx_co2mass.zip)

   CO2 was injected at specified surface condition(same as initial reservoir condition, 14432381.756 Pa and 368.15Kelvin) and well is under rate control(0.001m3/s, 5e6 seconds).  I'm trying to calculate the total CO2 dissolved in water. I remember Francois  mentioned that CompFraction isn't exported to vtk file due to some technical issue of 4d array, but I find globalCompFraction with values. What does this globalCompFraction  mean? Can I use this to calculate total dissolved CO2 mass?

   In test example, I tried another way to calculate:
   Total CO2 injected: 
   Injection rate* time* density = 0.001(m3/s)* 5e6(s)* 331.17(kg/m3)= 1655850 kg
   CO2 mass calculated from paraview: 
   ElementVolume* ReferencePorosoity* PhaseVolumeFraction_X* Fluid_PhaseMassDensity_X= 978267kg
   So the dissolved CO2 in water: = 1655850kg - 978267kg = 677583kg
   Is this correct? If no, can you advise me the right method to calculate the dissolved CO2 in GEOSX?

Thanks!

**Comment by francoishamon** [comment]
Hello Deming,
The simplest method to do what you are trying to do is the following:
- Checkout branch `bugfix/hamon/4DfieldsInVTK`
- Recompile
- Run your test
- In Paraview, you should be able to visualize `PhaseCompFraction_1`, which stores the phase component fractions (for CO2 and H2O components) in the brine phase. You can select the first component of `PhaseCompFraction_1`, and that will give you the amount of CO2 component dissolved in brine. Then you can compute the mass of CO2 dissolved in brine as `volume x porosity x phaseVolumeFraction_1 x phaseCompFraction_1_0 x phaseMassDensity_1`

**Reply by wdmhouston** [reply]
Francois,

   Thanks. Very useful and I will try!  This is the most critical output from CO2 simulation and I will verify all these values.


Regards,

Deming

**Reply by wdmhouston** [reply]
> Hello Deming,
> The simplest method to do what you are trying to do is the following:
> 
> * Checkout branch `bugfix/hamon/4DfieldsInVTK`
> * Recompile
> * Run your test
> * In Paraview, you should be able to visualize `PhaseCompFraction_1`, which stores the phase component fractions (for CO2 and H2O components) in the brine phase. You can select the first component of `PhaseCompFraction_1`, and that will give you the amount of CO2 component dissolved in brine. Then you can compute the mass of CO2 dissolved in brine as `volume x porosity x phaseVolumeFraction_1 x phaseCompFraction_1_0 x phaseMassDensity_1`

Francois,

    Thanks for your workaround. I was able to recompile GEOSX form that branch and got those components outputted.  I built a simple test case with 1 month injection. Here is the test case with all data included (results.pvsm is the paraview file with calculation steps)
[test_base.zip](https://github.com/GEOSX/GEOSX/files/6664286/test_base.zip)

In this simple case, CO2 is injected at constant surface condition with rate=0.0001m3/s and density: 646.9535807985 kg/m3(1619 psi/316.15K).
After one month(2592000 seconds), the expected total injection mass is  0.0001 m3/s * 646.9535807985 kg/m3 * 2592000 s =  167690.36814297 kg

In paraview, I calculated this way,
DissolvedCO2 = elementVolume*referencePorosity*phaseVolumeFraction_Y*fluid_phaseCompFraction_1_X*fluid_phaseMassDensity_Y
DissolvedCO2Total = sum(DissolvedCO2) = 57.9663 kg (@2592000 seconds)   
NonDissolvedCO2 = elementVolume*referencePorosity*fluid_phaseCompFraction_0_X*phaseVolumeFraction_X*fluid_phaseMassDensity_X
NonDissolvedCO2Total = sum(NonDissolvedCO2) = 438.21 kg (@2592000 seconds)
So the total CO2 calculated in Paraview is 496.1763 kg, and it's not close to 167690.36814297 kg.

Any idea the mistake in my calculation?

Thanks for your help with this calculation question.

**Reply by wdmhouston** [reply]
It looks like the fluid_phaseMassDensity_X outputted to Paraview is wrong. Please double check.

**Reply by francoishamon** [reply]
Hi Deming, could you please be more specific? Maybe post a screenshot of what you are seeing?

**Reply by wdmhouston** [reply]
![image](https://user-images.githubusercontent.com/19333848/123020847-94362c00-d398-11eb-8932-d570e8608a2c.png)
Here it is. I see the CO2 density of every cell is 354.429 kg/m3, which is lower than what I expected.

**Reply by francoishamon** [reply]
Hi Deming, 
In the example that you are posting, the pressure is equal to 1.4164e+08 Pa, which is way above the maximum pressure set in the table definitions for the CO2-brine EOS:
- co2flash.txt
- pvtliquid.txt
- pvtgas.txt

The maximum value is the second value that is written in each line: see the documentation, for instance  https://geosx-geosx.readthedocs-hosted.com/en/latest/coreComponents/constitutive/docs/CO2Brine.html#step-1-computation-of-the-phase-fractions-and-phase-component-fractions-flash

You currently use 1.5e7 Pa based on what I see in your files.

Also, you will see in the documentation that some elements of the CO2-brine model that we are using (specifically, the computation of the brine density and viscosity: https://geosx-geosx.readthedocs-hosted.com/en/latest/coreComponents/constitutive/docs/CO2Brine.html#brine-density-and-viscosity) are not valid above 5e7 Pa.

I suggest you try to lower the maximum BHP value for your injector to make sure that the pressure does not increase too much:
```
	<WellControls control="totalVolRate"
			              injectionStream="{ 1.0, 0 }"
			              name="wellControls"
			              referenceElevation="4.575"
			              targetBHP="5e7"
			              targetTotalRate="0.0001"
			              type="injector"
			              useSurfaceConditions="1"
			              surfacePressure="11162257.5929"
			              surfaceTemperature="316.15"/>
```
and then, in the CO2-brine PVT tables, you should set 5e7 Pa as the maximum value (not 1.5e7 Pa as in your current files)

**Reply by wdmhouston** [reply]
Francois, 

    You're right. I'll make a new test case and re-calculate all these values again.

**Reply by wdmhouston** [reply]
> Hi Deming,
> In the example that you are posting, the pressure is equal to 1.4164e+08 Pa, which is way above the maximum pressure set in the table definitions for the CO2-brine EOS:
> 
> * co2flash.txt
> * pvtliquid.txt
> * pvtgas.txt
> 
> The maximum value is the second value that is written in each line: see the documentation, for instance https://geosx-geosx.readthedocs-hosted.com/en/latest/coreComponents/constitutive/docs/CO2Brine.html#step-1-computation-of-the-phase-fractions-and-phase-component-fractions-flash
> 
> You currently use 1.5e7 Pa based on what I see in your files.
> 
> Also, you will see in the documentation that some elements of the CO2-brine model that we are using (specifically, the computation of the brine density and viscosity: https://geosx-geosx.readthedocs-hosted.com/en/latest/coreComponents/constitutive/docs/CO2Brine.html#brine-density-and-viscosity) are not valid above 5e7 Pa.
> 
> I suggest you try to lower the maximum BHP value for your injector to make sure that the pressure does not increase too much:
> 
> ```
> 	<WellControls control="totalVolRate"
> 			              injectionStream="{ 1.0, 0 }"
> 			              name="wellControls"
> 			              referenceElevation="4.575"
> 			              targetBHP="5e7"
> 			              targetTotalRate="0.0001"
> 			              type="injector"
> 			              useSurfaceConditions="1"
> 			              surfacePressure="11162257.5929"
> 			              surfaceTemperature="316.15"/>
> ```
> 
> and then, in the CO2-brine PVT tables, you should set 5e7 Pa as the maximum value (not 1.5e7 Pa as in your current files)

Francois,

   I updated this test model (lower injection rate and expand pressure upper limit in co2flash),
![image](https://user-images.githubusercontent.com/19333848/123660655-78e47a00-d7f9-11eb-91a2-56e0c7414092.png)
At the end of injection, the pressure goes to 2.8e7, but the fluid_phaseMassDensity_X is still 354. This number outputted to vtk seems never change.
here is the test case,
Can you check?


Regards,

Deming

**Reply by wdmhouston** [reply]
Francois, 

    I tried again with your suggestions, and I got correct density in place now. I will recalculate and validate the co2 mass in reservoir.

**Reply by francoishamon** [reply]
Ok great. If you checkout the latest version of `develop`, you will be able to output `phaseCompFraction` to vtk, which may be useful to compute the CO2 mass in the reservoir.

**Reply by wdmhouston** [reply]
Will try. Thanks!

**Reply by wdmhouston** [reply]
> Ok great. If you checkout the latest version of `develop`, you will be able to output `phaseCompFraction` to vtk, which may be useful to compute the CO2 mass in the reservoir.

Francois,
    In the latest develop version, do I have to change temperature in co2flash, pvtgas, pvtliquid from C to F?

**Reply by francoishamon** [reply]
Yes, there was an inconsistency in the units so now the temperature is expected in *Kelvin* in co2flash, pvtgas, and pvtliquid. 

In the base case you posted above, you have defined a temperature of 316.15 K in the solver
```
<CompositionalMultiphaseFVM discretization="fluidTPFA"
		                            fluidNames="{ fluid }"
		                            logLevel="2"
		                            maxCompFractionChange="0.2"
		                            name="compositionalMultiphaseFlow"
		                            relPermNames="{ relperm }"
		                            solidNames="{ rock }"
		                            targetRegions="{ Reservoir }"
		                            temperature="316.15"
		                            useMass="1"/>
```
so I suggest to use values consistent with that in the pvt files. For instance, for co2flash.txt you would have:
```
FlashModel CO2Solubility  1e6 5e7 5e4 315 317 1 0.4278
```
(for now the temperature is constant, we are are sure that is it will remain equal to 316.16 K everywhere during the entire simulation)

**Reply by wdmhouston** [reply]
Francois,

    Today, I re-tested the same case with temperature corrected to K unit in the latest develop branch version(see that attached test case files)
[CO2MassSimple.zip](https://github.com/GEOSX/GEOSX/files/6888774/CO2MassSimple.zip)

and here are the results of CO2 mass calculated from paraview and from injection rate.
    calculated from output properties in paraview:
         DissolvedCO2 in brine: 9972.36kg
         NonDissolvedCO2: 30690kg
         Total: 9972.36 + 30690 = 40662.36 kg
   
    calculated from injection rate:
            646.9535807985kg/m3 * 0.00002m3/s * 2592000s = 33538.07 kg.

    As we can see, the total CO2 mass doesn't match(no even close).  I've seen other cases with similar trend(over injected). Can you take a look and find out where the issue might be? If it's a manual calculation issue, please let me know which part is wrong. 


Thanks.


--




</body>

</html>


> Yes, there was an inconsistency in the units so now the temperature is expected in _Kelvin_ in co2flash, pvtgas, and pvtliquid.
> 
> In the base case you posted above, you have defined a temperature of 316.15 K in the solver
> 
> ```
> <CompositionalMultiphaseFVM discretization="fluidTPFA"
> 		                            fluidNames="{ fluid }"
> 		                            logLevel="2"
> 		                            maxCompFractionChange="0.2"
> 		                            name="compositionalMultiphaseFlow"
> 		                            relPermNames="{ relperm }"
> 		                            solidNames="{ rock }"
> 		                            targetRegions="{ Reservoir }"
> 		                            temperature="316.15"
> 		                            useMass="1"/>
> ```
> 
> so I suggest to use values consistent with that in the pvt files. For instance, for co2flash.txt you would have:
> 
> ```
> FlashModel CO2Solubility  1e6 5e7 5e4 315 317 1 0.4278
> ```
> 
> (for now the temperature is constant, we are are sure that is it will remain equal to 316.16 K everywhere during the entire simulation)

**Comment by francoishamon** [comment]
Hello @wdmhouston,
Thanks for the catch, there was indeed a problem in the injection rate set a surface condition. 
Now the problem should be fixed once #1536 is merged.

I attached your updated file for which I verified that the injection rate matches the amount of CO2 in the reservoir.
[deming.zip](https://github.com/GEOSX/GEOSX/files/6993433/deming.zip)

**Reply by wdmhouston** [reply]
Francois,

   Thanks. I tested in this simple case and I got about 32000kg. The calculated injection total is 33538.07. It's very close. Did you get similar results?

   I noticed there are some changes. One of the changes is :
<ConstantPermeability name="rockPerm"  permeabilityComponents="{ 1.0e-16, 1.0e-16, 1.0e-16 }"/>
  This is rock permeability and later it will be overwritten by permxFunc. Is my understanding correct?

> Hello @wdmhouston,
> Thanks for the catch, there was indeed a problem in the injection rate set a surface condition.
> Now the problem should be fixed once #1536 is merged.
> 
> I attached your updated file for which I verified that the injection rate matches the amount of CO2 in the reservoir.
> [deming.zip](https://github.com/GEOSX/GEOSX/files/6993433/deming.zip)

**Reply by francoishamon** [reply]
Ok, let me try again, I am a little bit surprised by your result. I think I was getting a much better match than that. Are you sure that you are using the final vtk file of the simulation?

**Reply by francoishamon** [reply]
@wdmhouston Here is what the simulator actually computes, checked by outputting values to the terminal:

- Total co2 mass injected =  co2Density * injectionVolRate * totalTime 
                                    = 647.006 kg.m^-3 * 0.00002 m3.s^-1 * 2,592,000 s = 33,541 kg.

- Total co2 mass in the reservoir = co2Saturation * co2PhaseDensity + co2InBrinePhaseCompFraction * brineSaturation * brinePhaseDensity = 33,541 kg.

(note: co2InCo2PhaseCompFraction = 1 in our PVT model)

So with the very tight Newton tolerance that I put in the xml I sent you (10^-9), we obtain a very good match.

The problem, if any, must be in the output. Can you please explain again how you grab the values to make your calculations? From the VTK file? At what time? Do you use the hdf5 output at all (I can show you how)?

**Comment by wdmhouston** [comment]
Francois,

    That is what I expected. Let me try again.


Regards,

Deming

On Tue, Aug 17, 2021 at 12:27 PM Francois Hamon ***@***.***>
wrote:

> @wdmhouston <https://github.com/wdmhouston> Here is what the simulator
> actually computes, checked by outputting values to the terminal:
>
>    -
>
>    Total co2 mass injected = co2Density * injectionVolRate * totalTime
>    = 647.006 kg.m^-3 * 0.00002 m3.s^-1 * 2592000 s = 33,541 kg.
>    -
>
>    Total co2 mass in the reservoir = co2Saturation * co2PhaseDensity +
>    co2InBrinePhaseCompFraction * brineSaturation * brinePhaseDensity = =
>    33,541 kg.
>
> (note: co2InCo2PhaseCompFraction = 1 in our PVT model)
>
> So with the very tight Newton tolerance that I put in the xml I sent you
> (10^-9), we obtain a very good match.
>
> The problem, if any, must be in the output. Can you please explain again
> how you grab the values to make your calculations? From the VTK file? At
> what time? Do you use the hdf5 output at all (I can show you how)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GEOSX/GEOSX/discussions/1457#discussioncomment-1196563>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AETQFWABQF4QPAWHLKIHDYTT5KLZJANCNFSM46OXKYRA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
Deming Wang
***@***.***

**Comment by wdmhouston** [comment]
Francois,

    I found an error in my old calculation(misunderstanding about co2
solubility unit in brine), and corrected it. Now the value is very close.

    I used that 1E-9 converge criteria and calculated in paraview this way:

    Calculator:
    DissolvedCO2 =
elementVolume*referencePorosity*phaseVolumeFraction_water*fluid_phaseCompFraction_water/co2*fluid_phaseMassDensity_water
    Python Calculator:
    DissolvedCO2Total = sum(DissolvedCO2) = 7949.6 kg

    CO2:
    Calculator:
    CO2 =
elementVolume*referencePorosity*phaseVolumeFraction_gas*fluid_phaseCompFraction_gas/co2*fluid_phaseMassDensity_gas
    Python Calculator:
    CO2Total = sum(NonDissolvedCO2) = 25403.8 kg

     Total CO2 in reservoir =  7949.6   kg + 25403.8 kg = 33353.4 kg.
 It's very close to 33,541 kg.


     You mentioned you calculated using hdf5 with 100% matched, can you
show me how to do it?


Regards,

Deming

On Tue, Aug 17, 2021 at 12:38 PM Deming Wang ***@***.***> wrote:

> Francois,
>
>     That is what I expected. Let me try again.
>
>
> Regards,
>
> Deming
>
> On Tue, Aug 17, 2021 at 12:27 PM Francois Hamon ***@***.***>
> wrote:
>
>> @wdmhouston <https://github.com/wdmhouston> Here is what the simulator
>> actually computes, checked by outputting values to the terminal:
>>
>>    -
>>
>>    Total co2 mass injected = co2Density * injectionVolRate * totalTime
>>    = 647.006 kg.m^-3 * 0.00002 m3.s^-1 * 2592000 s = 33,541 kg.
>>    -
>>
>>    Total co2 mass in the reservoir = co2Saturation * co2PhaseDensity +
>>    co2InBrinePhaseCompFraction * brineSaturation * brinePhaseDensity = =
>>    33,541 kg.
>>
>> (note: co2InCo2PhaseCompFraction = 1 in our PVT model)
>>
>> So with the very tight Newton tolerance that I put in the xml I sent you
>> (10^-9), we obtain a very good match.
>>
>> The problem, if any, must be in the output. Can you please explain again
>> how you grab the values to make your calculations? From the VTK file? At
>> what time? Do you use the hdf5 output at all (I can show you how)?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/GEOSX/GEOSX/discussions/1457#discussioncomment-1196563>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AETQFWABQF4QPAWHLKIHDYTT5KLZJANCNFSM46OXKYRA>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
>> .
>>
>
>
> --
> Deming Wang
> ***@***.***
>


-- 
Deming Wang
***@***.***
